File: clients/rust/src/linux.rs
use std::process::Command;

use crate::error::{KeySyncError, Result};
use crate::credential::CredentialEntry;
use crate::service::parse_service_name;

/// Retrieve a secret from libsecret via the `secret-tool` CLI.
pub(crate) fn get_secret(service: &str, account: &str) -> Result<String> {
    let output = Command::new("secret-tool")
        .args(["lookup", "service", service, "account", account])
        .output()?;

    if !output.status.success() {
        // secret-tool returns exit code 1 when the item is not found
        if output.status.code() == Some(1) {
            return Err(KeySyncError::NotFound);
        }
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(KeySyncError::KeychainError(
            format!("secret-tool lookup failed: {}", stderr.trim())
        ));
    }

    Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
}

/// List all keysync secrets by searching libsecret.
pub(crate) fn list_secrets() -> Result<Vec<CredentialEntry>> {
    let output = match Command::new("secret-tool")
        .args(["search", "service", "keysync"])
        .output()
    {
        Ok(o) => o,
        Err(_e) => return Ok(Vec::new()), // secret-tool not available, return empty
    };

    let out = String::from_utf8_lossy(&output.stdout);
    let lines: Vec<&str> = out.lines().collect();

    let mut entries = Vec::new();
    let mut current_svc = String::new();
    let mut current_acct = String::new();

    for line in &lines {
        let line = line.trim();

        if line.is_empty() {
            // End of a record block
            if !current_svc.is_empty() && !current_acct.is_empty() && current_svc.starts_with("keysync/") {
                let (scope, project) = parse_service_name(&current_svc);
                entries.push(CredentialEntry::new(scope, project, &current_acct));
            }
            current_svc.clear();
            current_acct.clear();
            continue;
        }

        if line.starts_with("service") {
            if let Some(val) = extract_attr(line) {
                current_svc = val;
            }
        } else if line.starts_with("account") {
            if let Some(val) = extract_attr(line) {
                current_acct = val;
            }
        }
    }

    // Handle last entry if no trailing blank line
    if !current_svc.is_empty() && !current_acct.is_empty() && current_svc.starts_with("keysync/") {
        let (scope, project) = parse_service_name(&current_svc);
        entries.push(CredentialEntry::new(scope, project, &current_acct));
    }

    Ok(entries)
}

/// Returns true if the error indicates the secret was not found.
pub(crate) fn is_not_found(err: &KeySyncError) -> bool {
    matches!(err, KeySyncError::NotFound)
}

/// Extract the value from an attribute line like "service = keysync/global"
fn extract_attr(line: &str) -> Option<String> {
    line.splitn(2, '=')
        .nth(1)
        .map(|v| v.trim().to_string())
}

File: clients/node/src/linux.ts
import { execFile } from "node:child_process";
import { KeySyncError } from "./index.js";

/** Linux keychain access via the `secret-tool` CLI (libsecret). */
export async function linuxGet(service: string, account: string): Promise<string> {
  try {
    const { stdout } = await execFilePromise("secret-tool", [
      "lookup", "service", service, "account", account,
    ]);
    const val = stdout.trim();
    if (!val) throw new KeySyncError("notFound", "secret not found");
    return val;
  } catch (err) {
    if (err instanceof KeySyncError) throw err;
    throw new KeySyncError("keychainError", `secret-tool lookup failed: ${err}`);
  }
}

/** List all keysync secrets on Linux by parsing `secret-tool search`. */
export async function linuxList(): Promise<Array<{ service: string; account: string }>> {
  let stdout: string;
  try {
    const result = await execFilePromise("secret-tool", [
      "search", "service", "keysync",
    ]);
    stdout = result.stdout;
  } catch {
    return []; // secret-tool not available or no results
  }

  const results: Array<{ service: string; account: string }> = [];
  let currentSvc = "";
  let currentAcct = "";

  for (const line of stdout.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) {
      if (currentSvc && currentAcct && currentSvc.startsWith("keysync/")) {
        results.push({ service: currentSvc, account: currentAcct });
      }
      currentSvc = "";
      currentAcct = "";
      continue;
    }
    if (trimmed.startsWith("service")) {
      currentSvc = parseAttrValue(trimmed);
    } else if (trimmed.startsWith("account")) {
      currentAcct = parseAttrValue(trimmed);
    }
  }
  // Handle last entry if no trailing blank line
  if (currentSvc && currentAcct && currentSvc.startsWith("keysync/")) {
    results.push({ service: currentSvc, account: currentAcct });
  }
  return results;
}

function parseAttrValue(line: string): string {
  const eqIdx = line.indexOf("=");
  return eqIdx >= 0 ? line.slice(eqIdx + 1).trim() : "";
}

function execFilePromise(
  cmd: string,
  args: string[]
): Promise<{ stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    execFile(cmd, args, (error, stdout, stderr) => {
      if (error) {
        reject(error);
      } else {
        resolve({ stdout, stderr });
      }
    });
  });
}
